logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-            return newRootType;
-        }
-
-        Resolvers resolvers() {
-            return resolvers;
-        }
-
-        HashMap<TypeVariable, Set<TypeVariable>> neighbours() {
-            return neighbours;
+        Traversal resolverTraversal() {
+            return resolverTraversal;
         }
 
-        private boolean hasNoInfo(TypeVariable variable) {
-            return (!variable.reference().isLabel() &&
-                    !variable.sub().isPresent() &&
-                    !variable.label().isPresent() &&
-                    variable.is().isEmpty());
+        Variable getVariable(Reference reference) {
+            return variableRegister.get(reference);
         }
 
-        private TypeVariable convert(Variable variable) {
-            if (variable.isType()) return convert(variable.asType());
-            else if (variable.isThing()) return convert(variable.asThing());
+        private void convert(Variable variable) {
+            if (variable.isType()) convert(variable.asType());
+            else if (variable.isThing()) convert(variable.asThing());
             else throw GraknException.of(ILLEGAL_STATE);
         }
 
         private TypeVariable convert(TypeVariable variable) {
-            if (resolvers.contains(variable)) return resolvers.resolver(variable);
-
-            TypeVariable resolver = resolvers.register(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
-
-            if (variable.constraints().isEmpty()) return resolver;
-
-            resolver = cloneVariable(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
-            return resolver;
+            if (resolverRegister.containsKey(variable.id())) return resolverRegister.get(variable.id());
+            resolverRegister.put(variable.id(), variable);
+            variableRegister.putIfAbsent(variable.reference(), variable);
+            variable.addTo(resolverTraversal);
+            return variable;
         }
 
         private TypeVariable convert(ThingVariable variable) {
-            if (resolvers.contains(variable)) return resolvers.resolver(variable);
-
-            TypeVariable resolver = resolvers.register(variable);
-            neighbours.putIfAbsent(resolver, new HashSet<>());
+            if (resolverRegister.containsKey(variable.id())) return resolverRegister.get(variable.id());
 
-            if (variable.constraints().isEmpty()) return resolver;
+            TypeVariable resolver = new TypeVariable(variable.reference().isAnonymous() ?
+                                                             newSystemId() : variable.id());
+            resolverRegister.put(variable.id(), resolver);
+            variableRegister.putIfAbsent(resolver.reference(), variable);
+            valueTypeRegister.putIfAbsent(resolver.id(), set());
 
-            variable.iid().ifPresent(constraint -> convertIID(resolver, constraint));
+            //Note: order is important!
```

Comment:
Added this comment because it's not obvious why we can't just iterate over `variable.constraints()`. The reason is that we're calling conversion on other variables we meet along the way (ie this a depth-search-first algorithm). And we have to make an assumption on `convertValue` that certain processes have occurred after calling convert(). So we have to put that one near the top.